pyomo/contrib/solver/solvers/knitro/utils.py

```diff
@@ -0,0 +1,272 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2025
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from collections.abc import Iterable, Mapping, MutableSet, Sequence
+from typing import Optional
+
+from pyomo.common.collections import ComponentMap, ComponentSet
+from pyomo.common.numeric_types import value
+from pyomo.contrib.solver.common.util import collect_vars_and_named_exprs
+from pyomo.contrib.solver.solvers.knitro.typing import Function
+from pyomo.core.base.block import BlockData
+from pyomo.core.base.constraint import Constraint, ConstraintData
+from pyomo.core.base.expression import Expression
+from pyomo.core.base.objective import Objective, ObjectiveData
+from pyomo.core.base.var import VarData
+from pyomo.core.expr.calculus.diff_with_pyomo import reverse_sd
+
+
+def get_active_objectives(block: BlockData) -> list[ObjectiveData]:
+    """Retrieve all active ObjectiveData objects from a Pyomo Block.
+
+    Args:
+        block (BlockData): The Pyomo block to search for objectives.
+
+    Returns:
+        list[ObjectiveData]: A sorted list of all active objectives in the block.
+
+    """
+    generator = block.component_data_objects(
+        Objective, descend_into=True, active=True, sort=True
+    )
+    return list(generator)
+
+
+def get_active_constraints(block: BlockData) -> list[ConstraintData]:
+    """Retrieve all active ConstraintData objects from a Pyomo Block.
+
+    Args:
+        block (BlockData): The Pyomo block to search for constraints.
+
+    Returns:
+        list[ConstraintData]: A sorted list of all active constraints in the block.
+
+    """
+    generator = block.component_data_objects(
+        Constraint, descend_into=True, active=True, sort=True
```

Comment:
It will be significantly more efficient to not use `sort=True` and instead use `sort=SortComponents.deterministic`